components/Editor/index.tsx

```diff
@@ -238,6 +238,83 @@ const Editor = ({ readOnly = false }: Props) => {
   ]
   useRegisterActions(actions, [highlightCode])
 
+  const handleCommentLine = useCallback(() => {
+    if (!editorRef.current) {
+      return
+    }
+    const textareaRef = editorRef.current._input
+    const lineNumberStart = textareaRef.value
+      .substring(0, textareaRef.selectionStart)
+      .split('\n').length
+    const lineNumberEnd = textareaRef.value
+      .substring(0, textareaRef.selectionEnd)
+      .split('\n').length
+
+    const selectionStart = textareaRef.selectionStart
+    const selectionEnd = textareaRef.selectionEnd
+    const lines = cairoCode.split('\n')
+    const linesToHandle: number[] = []
+    for (let k = lineNumberStart; k <= lineNumberEnd; k++) {
+      linesToHandle.push(k)
+    }
+
+    function isCommentLine(input: string) {
+      return input.startsWith('// ')
+    }
+
+    const isMultilineSelection = linesToHandle.length > 1
+    let charOffsetStart = 0
+    let charOffsetEnd = 0
+    if (isMultilineSelection) {
+      let commentedLines = 0
+      for (const lineNumber of linesToHandle) {
+        if (lines[lineNumber - 1] !== undefined) {
+          const line = lines[lineNumber - 1]
+          if (isCommentLine(line)) {
+            lines[lineNumber - 1] = line.substring(3)
+          } else {
+            lines[lineNumber - 1] = '// ' + line
+            commentedLines += 1
+          }
+        }
+      }
+      charOffsetEnd = commentedLines * 3
+    } else {
+      const lineNumber = linesToHandle[0]
+      const line = lines[lineNumber - 1]
+      if (isCommentLine(line)) {
+        lines[lineNumber - 1] = line.substring(3)
+        charOffsetStart = -3
+        charOffsetEnd = -3
+      } else {
+        lines[lineNumber - 1] = '// ' + line
+        charOffsetStart = 3
+        charOffsetEnd = 3
+      }
+    }
+
+    setCairoCode(lines.join('\n'))
+    setTimeout(() => {
```

Comment:
we should clear the timeout by calling `clearTimeout()` in the hook cleanup, so that it doesn't hang in memory

I found this article which might serve as an inspiration: https://borstch.com/snippet/handling-cleanup-in-usecallback-memoization

Reply:
Added the cleanup + minor fix for the commenting logic for some edge cases.

Reply:
I read the article and it appeared strange that there is a cleanup logic implemented as a function returned from `useCallback`. I tested the solution and it seemed that the cleanup function wasn't called at all. Because of this I decided to implement this using ref which seems to be working as expected. If you have some feedback on this, please let me know.